Updated Evaluation (#4)
liuhenry4428/cs6120: content/blog/2019-09-25-RecordTypes.md

Immutable data structures are easier to reason about when developing compiler optimizations. However, forcing developers to recreate the entire record every time they need to change a value poses usability challenges. We decided on sticking to immutability in part for optimizations' sake. The ability to make guarantees that a value will not change lends itself well to constant-folding, even after a function call that takes in your record as an argument. 

### Evaluation
To evaluate our implementation, we need to check that our implementation works as specified as a language feature, and that this language features satisfies the goal of extending Bril's ability to compile higher-level languages with record data types.

One of the goals of record types was to allow Bril to logically group and use related data points. This feature is useful when compiling higher-level languages that utilize a similar data structure like records in OCaml or structs in C into Bril. 

> To measure success of this goal, we considered whether the functionality added from these new types increases Bril's ability to compile higher-level languages with record type data structures.

Reviewer: I think this may be a little redundant as 

> ...and that this language features satisfies the goal of extending Bril's ability to compile higher-level languages with record data types.

 was mentioned right above

Reviewer 2: yea i also was thinking that. Maybe we say "to measure the success of the latter goal, ..."

Reviewer 2: and delete the first paragraph

We implemented immutable nominal record types. This implementation includes record declaration, instantiation, and access. Decisions about which operations to include in our design were influenced by the operations available on record data types in higher-level languages. To test this functionality, we created a suite of tests that covered each of these operations, as well as combinations of these operations.

The primary operation not supported by our record type specification that is supported by some higher-level language record types is mutation. While our record types do not support mutation as discussed, this does not significantly hinder Bril's ability to compile higher-level languages that support this feature. 

Consider the following C code that declares a struct and provides a method to update.
```
struct Person {
   int age;
   bool isAsleep;
};

void updateAge(Person p){
    p->age = p->age + 1;
}

int main(void) {
    Person Henry = { 20, false };

>     updateAge(Henry);

Reviewer: ```suggestion
    Henry.age += 1;
```
Can we inline the function to make it more similar to the bril?

Reviewer 2: yea for sure!

    return 0;
}
```
Compiling this program to Bril may look something like this, assuming we ignore function calls.
```
type Person = {age: int; isAsleep: bool};
v0: int = const 20;
v1: bool = const false;
Henry: Person = record {age: v0; isAsleep: v1};
v2: int = const 21;
Henry: Person = Henry with {age: v2};
```

> As shown in this example, mutable record types can be compiled into immutable records in Bril without significant effort. Therefore, lack of this operation does not compromise this goal.

Reviewer: ```suggestion
As shown in this example, mutable record types can be transformed into immutable records in Bril without significant effort. Therefore, lack of this operation does not compromise this goal.
```
Can we say "transformed" instead to make it clear that a compiler's not involved?

Reviewer 2: yep!

Finally, to evaluate record types as a language feature in Bril, we consider how this functionality translates to Bril.
We found that creating new records was a tedious process if the record was large, so we implemented *with* statements in addition to the features mentioned above for situations where one wanted to duplicate a record with a few changes. It should be noted that it is bad form to use a with statement with no fields because that would be identical to referencing the old record with `… = id oldRecordName`.

We were successful in this aspect as creating a new record from an existing one with this syntax is more concise and easier to reason about than copying over every field. This is an advantage in an IR as we can logically think about a single _with statement_ and a sequence of functionally equivalent statements that copy variables in the same way.

Consider the blocks of code below. These programs duplicate a record with one field changed.
Here we show what this would look like without _with statements_. 

```

> type Person = {age: int; isHappy: bool};

Reviewer 2: ```suggestion
type Person = {age: int; isAsleep: bool};
```

v0: int = const 21;
v1: bool = const false;

> Henry: Person = record {age: v0; isHappy: v1};

Reviewer 2: ```suggestion
Henry: Person = record {age: v0; isAsleep: v1};
```

v2: bool = const true;
v3: int = Henry . age;
AwakeHenry: Person = record {age: v3; isAsleep: v2};
```

> Here we use the with syntax.

Reviewer: ```suggestion
Here we use the *with* syntax.
```

Reviewer 2: _with_?

Reviewer 2: ah i see, yes i agree.

```
type Person = {age: int; isAsleep: bool};
v0: int = const 21;
v1: bool = const false;

> Henry: Person = record {age: v0; isAsleep: v1};

Reviewer: Can we keep the examples consistent with either isHappy or isAsleep?

Reviewer 2: oh shit yes didnt see that

Reviewer 2: lets do asleep

v2: bool = const true;
AwakeHenry: Person = Henry with {isAsleep: v2};
```

> It is worth noting that the size of code required to duplicate a record without _with statements_ scales linearly with the size of the record. In contrast, the size of code required to duplicate a record with _with statements_ does not change as the size of the record increases. Therefore, record types are successsful as a language feature as they integrate well with current syntax and do not impose unneccesary code bloat. 

Reviewer: ```suggestion
It is worth noting that the size of code required to duplicate a record without _with statements_ scales linearly with the size of the record. In contrast, the size of code required to duplicate a record with _with statements_ only increases with the size of the changes. Therefore, record types are successful as a language feature as they integrate well with current syntax and do not impose unnecessary code bloat. 
```
I think it's important to note that the size of the with instruction is not technically unchanged as one would have to write more field-value-pairs if they want more changes. This would also be reflected into whatever assembly Bril is compiled into.

Reviewer 2: hmm both with statements and without with statements scale in that way as well. Maybe we can mention they are the same in that regard, but the difference is that using with statements does not change in size as you increase number of record fields

> Overall, record types implement the specified record operations, and increase Bril's ability to compiler higher-level languages. 

Reviewer: ```suggestion
Overall, record types implement the basic operations necessary to use them effectively and they increase Bril's ability to compile higher-level languages. 
```

Reviewer 2: yes
